GitExternal.cmake

```diff
@@ -101,67 +109,68 @@ function(GIT_EXTERNAL DIR REPO TAG)
-    return()
-  endif()
-
-  # update to given tag
-  execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse --short HEAD
-    OUTPUT_VARIABLE currentref OUTPUT_STRIP_TRAILING_WHITESPACE
-    WORKING_DIRECTORY ${DIR})
-  git_external_message(
-    "current ref is \"${currentref}\" and tag is \"${TAG}\"")
-  if(currentref STREQUAL TAG) # nothing to do
-    return()
-  endif()
-
-  # reset generated files
-  foreach(GIT_EXTERNAL_RESET_FILE ${GIT_EXTERNAL_RESET})
-    git_external_message("git reset -q ${GIT_EXTERNAL_RESET_FILE}")
-    execute_process(
-      COMMAND "${GIT_EXECUTABLE}" reset -q "${GIT_EXTERNAL_RESET_FILE}"
-      RESULT_VARIABLE nok ERROR_VARIABLE error
-      WORKING_DIRECTORY "${DIR}")
-    git_external_message("git checkout -q -- ${GIT_EXTERNAL_RESET_FILE}")
-    execute_process(
-      COMMAND "${GIT_EXECUTABLE}" checkout -q -- "${GIT_EXTERNAL_RESET_FILE}"
-      RESULT_VARIABLE nok ERROR_VARIABLE error
-      WORKING_DIRECTORY "${DIR}")
-  endforeach()
-
-  # fetch latest update
-  execute_process(COMMAND "${GIT_EXECUTABLE}" fetch origin -q
-    RESULT_VARIABLE nok ERROR_VARIABLE error
-    WORKING_DIRECTORY "${DIR}")
-  if(nok)
-    message(STATUS "Update of ${DIR} failed:\n   ${error}")
-  endif()
-
-  # update tag
-  git_external_message("git rebase FETCH_HEAD")
-  execute_process(COMMAND ${GIT_EXECUTABLE} rebase FETCH_HEAD
-    RESULT_VARIABLE RESULT ERROR_QUIET OUTPUT_QUIET
-    WORKING_DIRECTORY "${DIR}")
-  if(RESULT)
-    message(STATUS "git rebase failed, aborting ${DIR} merge")
-    execute_process(COMMAND ${GIT_EXECUTABLE} rebase --abort
-      WORKING_DIRECTORY "${DIR}" ERROR_QUIET OUTPUT_QUIET)
-  endif()
+  file(RELATIVE_PATH __dir ${CMAKE_SOURCE_DIR} ${DIR})
+  string(REGEX REPLACE "[:/]" "-" __target "${__dir}")
+  set(__rebase_cmake "${CMAKE_CURRENT_BINARY_DIR}/${__target}-rebase.cmake")
+  file(WRITE ${__rebase_cmake}
+    "if(NOT IS_DIRECTORY ${DIR}/.git)\n"
+    "  message(FATAL_ERROR \"Can't update git external ${__dir}: Not a git repository\")\n"
+    "endif()\n"
+    # check if we are already on the requested tag
+    "execute_process(COMMAND ${GIT_EXECUTABLE} rev-parse --short HEAD\n"
+    "  OUTPUT_VARIABLE currentref OUTPUT_STRIP_TRAILING_WHITESPACE\n"
+    "  WORKING_DIRECTORY ${DIR})\n"
+    "if(currentref STREQUAL ${TAG}) # nothing to do\n"
+    "  return()\n"
+    "endif()\n"
+    "\n"
+    # reset generated files
+    "foreach(GIT_EXTERNAL_RESET_FILE ${GIT_EXTERNAL_RESET})\n"
+    "  execute_process(\n"
+    "    COMMAND \"${GIT_EXECUTABLE}\" reset -q \"\${GIT_EXTERNAL_RESET_FILE}\"\n"
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "  execute_process(\n"
+    "    COMMAND \"${GIT_EXECUTABLE}\" checkout -q -- \"${GIT_EXTERNAL_RESET_FILE}\"\n"
+    "    ERROR_QUIET OUTPUT_QUIET\n"
+    "    WORKING_DIRECTORY \"${DIR}\")\n"
+    "endforeach()\n"
+    "\n"
+    # fetch latest update
+    "execute_process(COMMAND \"${GIT_EXECUTABLE}\" fetch origin -q\n"
```

Comment:
make 'origin' a configurable variable? I'd like to fetch from 'user' or even '-all' sometimes.